protocol-units/dispute/src/RStarM.sol

```diff
@@ -0,0 +1,259 @@
+        uint256 _m, 
+        uint256 control_id_0, 
+        uint256 control_id_1,
+        uint256 bn254_control_id
+    ) {
+        delta = _delta;
+        p = _p;
+        m = _m;
+        CONTROL_ID_0 = control_id_0;
+        CONTROL_ID_1 = control_id_1;
+        BN254_CONTROL_ID = bn254_control_id;
+    }
+
+    /// @notice splits a digest into two 128-bit words to use as public signal inputs.
+    /// @dev RISC Zero's Circom verifier circuit takes each of two hash digests in two 128-bit
+    /// chunks. These values can be derived from the digest by splitting the digest in half and
+    /// then reversing the bytes of each.
+    function splitDigest(bytes32 digest) internal pure returns (uint256, uint256) {
+        uint256 reversed = reverseByteOrderUint256(uint256(digest));
+        return (uint256(uint128(uint256(reversed))), uint256(reversed >> 128));
+    }
+
+    function registerValidator() external payable {
+        require(msg.value >= MIN_STAKE, "Insufficient stake");
+        require(!validators[msg.sender].isRegistered, "Validator already registered");
+        validators[msg.sender] = Validator(true, msg.value);
+        emit ValidatorRegistered(msg.sender, msg.value);
+    }
+
+    function getValidator(address validator) external view returns (bool, uint256) {
+        return (validators[validator].isRegistered, validators[validator].stake);
+    }
+
+    function deregisterValidator() external {
+        Validator storage validator = validators[msg.sender];
+        require(validator.isRegistered, "Validator not registered");
+        require(validator.stake > 0, "No stake to withdraw");
+        uint256 stake = validator.stake;
+        validator.isRegistered = false;
+        validator.stake = 0;
+        payable(msg.sender).transfer(stake);
+        emit ValidatorDeregistered(msg.sender);
+    }
+
+    // We probably want to use bytes calldata here to reduce gas but going with bytes32 for now as we 
+    // need it elsewhere.
+    function submitDispute(bytes32 blockHash) external {
+        bytes32 disputeHash = keccak256(abi.encodePacked(blockHash, msg.sender, block.timestamp));
+        require(disputes[disputeHash].timestamp == 0, "Dispute already submitted");
+        disputes[disputeHash] = Dispute(blockHash, block.timestamp, DisputeState.SUBMITTED);
+        emit DisputeSubmitted(disputeHash, blockHash, msg.sender);
+    }
+
+    function resolveDispute(bytes32 disputeHash, DisputeState state) external {
+        require(state > DisputeState.SUBMITTED && state <= DisputeState.UNVERIFIABLE, "Invalid state transition");
+        Dispute storage dispute = disputes[disputeHash];
+        require(dispute.timestamp != 0, "Dispute not found");
+        require(dispute.state == DisputeState.SUBMITTED, "Dispute already resolved");
+        dispute.state = state;
+        emit DisputeResolved(disputeHash, state);
+    }
+
+    function submitProof(bytes32 blockHash, Receipt calldata receipt) external {
```

Comment:
So, still for RStarM we would want those optimistic commitment rounds. But, hold off on patching that in for a bit.